Add an XRT plugin

Ralender opened the pull request:
This is based on top of https://github.com/triSYCL/sycl/pull/162.

The plugin is not complete. This is still a work in progress.

Commit:
[XRT][PI][SYCL] Add initial XRT pluging

Commit:
[XRT][SYCL] Add XRT PI backend slection

Commit:
[XRT][PI] Can execute basic complete examples

keryell commented:
To be rebased somehow on `unified/next`?

Ralender commented:
> To be rebased somehow on `unified/next`?

I rebased it but this is still a draft.

keryell reviewed (commented):
Quite more readable!
Thanks.

keryell commented:
Can you push your latest version?

Ralender commented:
Pushed but still experimental.

Commit:
[PI][XRT] Add implementation for xrt_piEnqueueMemBufferWrite, xrt_piQueueFinish, xrt_piQueueFlush

Also bypass a few XRT weirdness. and add the reproducer system

Commit:
[SYCL][Vitis] Update documentation with new changes to tests and the new XRT backend

Commit:
[XRT][Vitis][Test] Refactor test parallelisme, further improve xrt_pi, terminate xsimk

 - refactor test parallelism allowing test in hw_emu to run in ~2500s from ~8500s
 - add features to the xrt_pi, finally all test should pass.
 - add feature to terminate the simluator process

Ralender commented:
now all test pass with the xrt plugin so it is probably not experimental anymore

keryell commented on sycl/include/CL/sycl/detail/terminate_xsimk.hpp:
```diff
@@ -0,0 +1,76 @@
+//==------------ sycl_mem_obj_t.hpp - SYCL standard header file ------------==//
+//
```
There is a need for a big comment explaining the mess here. :-)

keryell commented on sycl/include/CL/sycl/detail/terminate_xsimk.hpp:
```diff
@@ -0,0 +1,76 @@
+
+#pragma once
+
+#ifndef __linux__
```
? Comment.

keryell commented on sycl/include/CL/sycl/detail/terminate_xsimk.hpp:
```diff
@@ -0,0 +1,76 @@
+#include <fstream>
+#include <vector>
+#include <cassert>
+#include <signal.h>
```
Sort.

keryell commented on sycl/include/CL/sycl/detail/terminate_xsimk.hpp:
```diff
@@ -0,0 +1,76 @@
+#include <signal.h>
+
+inline void terminate_xsimk() {
+  int my_pid = ::getpid();
```
```c++
     #include <sys/types.h>
     #include <unistd.h>
```
somewhere?

keryell replied:
And of course it is not an `int`.

keryell commented on sycl/include/CL/sycl/detail/terminate_xsimk.hpp:
```diff
@@ -0,0 +1,76 @@
+  std::unordered_map<int, std::vector<int>> ppid2pids;
+  /// set of xsimk processes
+  std::unordered_set<int> xsimks;
+  std::filesystem::path proc("/proc/");
```
```suggestion
  std::filesystem::path proc("/proc");
```
?

keryell commented on sycl/include/CL/sycl/detail/terminate_xsimk.hpp:
```diff
@@ -0,0 +1,76 @@
+    std::string s = entry.stem();
+    if (s.empty() || !(s.find_first_not_of("0123456789") == std::string::npos))
+      continue;
+    int pid = std::stoi(s);
```
If we use exceptions you could just `try`/`catch` this an remove lines 37 and 38.

keryell commented on sycl/include/CL/sycl/detail/terminate_xsimk.hpp:
```diff
@@ -0,0 +1,76 @@
+    int pid = std::stoi(s);
+    std::ifstream file(entry / "stat");
+    if (!file.is_open())
+      continue;
```
Add comment. For example the process has stopped already.

keryell replied:
Actually you can remove it if you use a `try`/`catch`.

keryell commented on sycl/include/CL/sycl/detail/terminate_xsimk.hpp:
```diff
@@ -0,0 +1,76 @@
+      continue;
+    int ppid = std::stoi(words[3]);
+    if (words[1] == "(xsimk)")
+      xsimks.insert(pid);
```
Cryptic. Express directly the intent according to `proc` man page:
```suggestion
   file.exceptions(std::ifstream::failbit);
   try {
     pid_t pid;
     file >> pid;
     std::string short_command_name;
     file >> short_command_name;
     std::string stat;
     file >> stat;
     pid_t parent_pid;
     file >> parent_pid;
     if (short_command_name == "(xsimk)")
       xsimks.insert(pid);
       ...
       } catch (const std::ios_base::failure& fail) {
...
    }
```

keryell commented on sycl/include/CL/sycl/detail/terminate_xsimk.hpp:
```diff
@@ -0,0 +1,76 @@
+  int my_pid = ::getpid();
+
+  /// tree of processes
+  std::unordered_map<int, std::vector<int>> ppid2pids;
```
Use the right types
```suggestion
  std::unordered_map<pid_t, std::vector<pid_t>> ppid2pids;
```

keryell commented on sycl/include/CL/sycl/detail/terminate_xsimk.hpp:
```diff
@@ -0,0 +1,76 @@
+  /// tree of processes
+  std::unordered_map<int, std::vector<int>> ppid2pids;
+  /// set of xsimk processes
+  std::unordered_set<int> xsimks;
```
```suggestion
  std::unordered_set<pid_t> xsimks;
```

keryell commented on sycl/include/CL/sycl/detail/terminate_xsimk.hpp:
```diff
@@ -0,0 +1,76 @@
+inline void terminate_xsimk() {
+  int my_pid = ::getpid();
+
+  /// tree of processes
```
I have some troubles to name this :-)
```suggestion
  /// Tree of processes according to their parenthood
```

keryell commented on sycl/include/CL/sycl/detail/terminate_xsimk.hpp:
```diff
@@ -0,0 +1,76 @@
+    int pid = stack.back();
+    stack.pop_back();
+    if (xsimks.count(pid))
+      ::kill(pid, 9);
```
No magic number
```c++
#include <signal.h>
#include <sys/types.h>
::kill(pid, SIGKILL);
```

keryell commented on sycl/plugins/opencl/pi_opencl.cpp:
```diff
@@ -1397,6 +1398,7 @@ pi_result piextKernelGetNativeHandle(pi_kernel kernel,
 // pi_level_zero.cpp for reference) Currently this is just a NOOP.
 pi_result piTearDown(void *PluginParameter) {
   (void)PluginParameter;
+  terminate_xsimk();
```
Comment

keryell replied:
Comment++

keryell replied:
HLD simulator?

keryell replied:
Ping?

keryell commented on sycl/include/CL/sycl/detail/terminate_xsimk.hpp:
```diff
@@ -0,0 +1,76 @@
+    ppid2pids[ppid].push_back(pid);
+  }
+
+  /// traverse the tree from the current process to terminate xsimks
```
```suggestion
  /// traverse the tree from the current process to terminate our xsimks
```

keryell commented on sycl/plugins/opencl/pi_opencl.cpp:
```diff
@@ -1408,6 +1410,8 @@ pi_result piPluginInit(pi_plugin *PluginInit) {
     return PI_INVALID_OPERATION;
   }
 
+  std::atexit(terminate_xsimk);
```
Comment

keryell commented on sycl/plugins/xrt/CMakeLists.txt:
```diff
@@ -0,0 +1,68 @@
+    message(STATUS "Standard XRT installation path found")
+    set(XILINX_RT_INCLUDE "/opt/xilinx/xrt/include")
+  
+  #TODO: Automatic download of XRT once their CMake is clean
```
Interesting feature. Do you mean a FETCH_CONTENT for example?
But let's not over-engineering it for now :-)

keryell commented on sycl/plugins/xrt/CMakeLists.txt:
```diff
@@ -0,0 +1,68 @@
+  )
+
+  if (MSVC)
+    # by defining __SYCL_BUILD_SYCL_DLL, we can use __declspec(dllexport)
```
Start sentences with uppercase
```suggestion
    # By defining __SYCL_BUILD_SYCL_DLL, we can use __declspec(dllexport)
```
and recurse though the PR :-)

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2222 @@
+#endif
+
+/// XRT has a bunch of very noisy warnings we need to suppress.
+/// XRT can be found as opencl implementation so we suppress warning on both xrt
+/// native header and opencl headers
```
```suggestion
/// XRT can be found as OpenCL implementation so we suppress warning on both XRT
/// native header and OpenCL headers
```

keryell replied:
Really? Why do these warnings not appear otherwise?
@stsoe ?

Ralender replied:
the warning likely come from us using a very recent compiler. just to be clear XRT's OpenCL headers are included not by XRT headers by by the pi headers.

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2222 @@
+#include <optional>
+#include <stdint.h>
+#include <string>
+#include <vector>
```
Sooooooort.

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2222 @@
+
+#pragma clang diagnostic pop
+
+// Most of the API is not implemented and exist so most parameters are not used
```
I do not understand the sentence with "and exist".

Ralender replied:
what i mean is that unimplented function look like:
```
pi_result xrt_piEnqueueMemBufferMap(pi_queue command_queue, pi_mem buffer,
                                    pi_bool blocking_map,
                                    pi_map_flags map_flags, size_t offset,
                                    size_t size,
                                    pi_uint32 num_events_in_wait_list,
                                    const pi_event *event_wait_list,
                                    pi_event *event, void **ret_map) {
  sycl::detail::pi::unimplemented(__PRETTY_FUNCTION__);
}
```
so the parameters are all going to be emit warnings.

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2222 @@
+#pragma clang diagnostic ignored "-Wunused-parameter"
+#pragma clang diagnostic ignored "-Wgnu-zero-variadic-macro-arguments"
+
+/// Base class for _pi_* object that need ref-counting
```
```suggestion
/// Base class for _pi_* objects that need ref-counting
```

keryell reviewed (changes requested):
Great massive work.
Stopping here for today. :-)

Commit:
[PI][XRT] Address comments, reformat

Commit:
[XRT][PI] Refactor ref_counting, start adding interop support to the xrt pi

Commit:
[SYCL][PI] go throught the device cache even for sycl::device created from an interop handle

keryell commented on sycl/include/CL/sycl/backend.hpp:
```diff
@@ -23,6 +23,11 @@
+#if SYCL_BACKEND_XRT
+#include <CL/sycl/detail/backend_traits_xrt.hpp>
+#else
+#error "meh
```
Gnih?

keryell commented on sycl/include/CL/sycl/detail/backend_traits_xrt.hpp:
```diff
@@ -0,0 +1,120 @@
+namespace xrt {
+struct device;
+struct xclbin;
+struct bo;
```
Sort.
But just curious, since there is no cyclic dependency with us, why not just including `xrt.h` since it will be included at some point anyway to be used by the end-user?

Ralender replied:
there is no cyclic dependencies. but the XRT includes path is not configured by the CMake of the sycl runtime (only the pi_xrt) while compiling this file.

keryell replied:
Add this as a comment.

keryell replied:
At least `bo` can be first

keryell commented on sycl/include/CL/sycl/feature_test.hpp.in:
```diff
@@ -82,6 +82,10 @@ namespace sycl {
 #if SYCL_BUILD_PI_HIP
 #define SYCL_EXT_ONEAPI_BACKEND_HIP 1
 #endif
+#cmakedefine01 SYCL_BUILD_PI_XRT
```
Gnih?

Ralender replied:
this is not a C++ file yet. it will be propocessed by cmake into a C++ header. with
either
#define SYCL_BUILD_PI_XRT 1 or #define SYCL_BUILD_PI_XRT 0

keryell commented on sycl/plugins/xrt/reproducer.h:
```diff
@@ -0,0 +1,367 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
```
Explain what this is for with a marvelous story

keryell commented on sycl/plugins/xrt/reproducer.h:
```diff
@@ -0,0 +1,367 @@
+#endif
+}
+
+struct RawTypeNameFormat {
```
Explain all these atrocities. At some point I wonder whether it is not easier to just use Boost or https://github.com/Neargye/nameof as a dependency.

Ralender replied:
I don't think it is worth the dependency, or worth explaining.

keryell commented on sycl/plugins/xrt/reproducer.h:
```diff
@@ -0,0 +1,367 @@
+    std::string id = "name" + std::to_string(id_next++);
+    int e = std::snprintf(nullptr, 0, format.c_str(), id.c_str());
+    name.resize(e + 1);
+    std::snprintf(&name[0], name.size(), format.c_str(), id.c_str());
```
I am vomiting here :-)

Ralender replied:
we dont have std::format yet and I needed formatting. so there is nothing better.

and yes I agree it is horrible to have to deal with std::string being counted but not null-terminated and snprintf adding null-terminators.

keryell replied:
We can have `std::format` with C++20! You know? The last version of C++...

keryell commented on sycl/plugins/xrt/reproducer.h:
```diff
@@ -0,0 +1,367 @@
+constexpr std::true_type is_streamable_impl(T) {
+  return {};
+}
+constexpr std::false_type is_streamable_impl(...) { return {}; }
```
Variadic arguments for ever!

keryell commented on sycl/plugins/xrt/reproducer.h:
```diff
@@ -0,0 +1,367 @@
+          typename std::enable_if_t<
+              std::is_same_v<
+                  void, std::void_t<decltype(std::declval<T>().get_handle())>>,
+              int> = 0>
```
But whyyyyy?
All this is new code, we compile in C++20, so just use concepts.

Ralender replied:
no we dont compile in C++20 but in C++17. because it is very restrictive to expect users to have a C++20 compiler.

keryell commented on sycl/plugins/xrt/reproducer.h:
```diff
@@ -0,0 +1,367 @@
+constexpr std::false_type has_get_handle_impl(...) { return {}; }
+
+template <typename T>
+using remove_cvref_t = std::remove_reference_t<std::remove_cv_t<T>>;
```
C++20 is good for your health

keryell commented on sycl/tools/sycl-vxx/bin/sycl_vxx.py:
```diff
@@ -181,7 +183,43 @@ def _run_in_isolated_proctree(cmd, *args, **kwargs):
+    # each instance of vxx uses 4 gb at most and we keep 10% margin
+    max_vxx_instance_count = int(math.trunc(ram_gb * 0.9) / 4)
+    if max_vxx_instance_count == 0:
+        print("warning: v++ is likely to run out of ram")
```
```suggestion
        print("warning: v++ is likely to run out of RAM")
```

keryell commented on sycl/tools/sycl-vxx/bin/sycl_vxx.py:
```diff
@@ -181,7 +183,43 @@ def _run_in_isolated_proctree(cmd, *args, **kwargs):
+                # kb to gb
+                ram_gb = int(line.split()[1]) / (1024 * 1024)
+                break
+    # each instance of vxx uses 4 gb at most and we keep 10% margin
```
```suggestion
    # each instance of vxx uses 4 GiB at most and we keep 10% margin
```

keryell commented on sycl/tools/sycl-vxx/bin/sycl_vxx.py:
```diff
@@ -181,7 +183,43 @@ def _run_in_isolated_proctree(cmd, *args, **kwargs):
+    with open('/proc/meminfo') as file:
+        for line in file:
+            if 'MemAvailable' in line:
+                # kb to gb
```
```suggestion
                # KiB to GiB
```

keryell reviewed (changes requested):
Interesting

Ralender committed:
Apply suggestions from code review

Co-authored-by: Ronan Keryell <ronan@keryell.fr>

Commit:
[SYCL][PI][XRT] Refactor pi_xrt's ownership management, add support for more interop API

Commit:
Address review comments

Commit:
[SYCL][PI][XRT] Finish interop for kernels, cleanup xrt_pi

keryell commented on sycl/include/CL/sycl/detail/backend_traits_xrt.hpp:
```diff
@@ -0,0 +1,84 @@
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
```
```suggestion
namespace sycl::detail {
```

keryell commented on sycl/include/CL/sycl/detail/terminate_xsimk.hpp:
```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+/// The implementation of terminate_xsimk depends on many linux or posix
+/// specific features so it is only provided on linux
```
Use the right spelling for LiNUx and PoSIx

keryell commented on sycl/include/sycl/ext/xilinx/xrt.hpp:
```diff
@@ -0,0 +1,127 @@
+#include <CL/sycl/backend.hpp>
+#include <CL/sycl/program.hpp>
+#include <CL/sycl/detail/pi.h>
+#include <CL/sycl/backend.hpp>
```
`std::sort`, `std::unique` :-)

keryell commented on sycl/include/sycl/ext/xilinx/xrt.hpp:
```diff
@@ -0,0 +1,127 @@
+#include <xrt.h>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
```
Do not tell me you want C++14 :-)
```suggestion
namespace sycl::detail {
```

keryell replied:
Ah yes, I was wrong here on the reason. You really 2 scopes.

keryell commented on sycl/include/sycl/ext/xilinx/xrt.hpp:
```diff
@@ -0,0 +1,127 @@
+template <typename To>
+typename std::enable_if<!std::is_reference_v<To>, To>::type
+from_native_handle(pi_native_handle handle) {
+  return (To)(handle);
```
```suggestion
  return { handle };
```
?

keryell commented on sycl/include/sycl/ext/xilinx/xrt.hpp:
```diff
@@ -0,0 +1,127 @@
+/// The backend is responsible for selecting the correct type and reference
+/// kind based on the semantic of the underlying pi call.
+template <typename To>
+typename std::enable_if<!std::is_reference_v<To>, To>::type
```
```suggestion
typename std::enable_if_t<!std::is_reference_v<To>, To>
```

keryell commented on sycl/include/sycl/ext/xilinx/xrt.hpp:
```diff
@@ -0,0 +1,127 @@
+  /// what it receives
+  return std::forward<To>(
+      *reinterpret_cast<std::remove_reference_t<To> *>(handle));
+}
```
Syntax error. Why `std::forward<To>`?

Ralender replied:
What is the syntax error ?
the std::forward if optional here since the return is going to do the cast if need anyway, but it indicates to the reader that To could be either kind of reference.

keryell commented on sycl/include/sycl/ext/xilinx/xrt.hpp:
```diff
@@ -0,0 +1,127 @@
+}
+
+template <typename To>
+typename std::enable_if<std::is_reference_v<To>, To>::type
```
```suggestion
typename std::enable_if_t<std::is_reference_v<To>, To>
```

keryell replied:
Everywhere

keryell commented on sycl/include/sycl/ext/xilinx/xrt.hpp:
```diff
@@ -0,0 +1,127 @@
+template <typename From>
+typename std::enable_if<!std::is_reference_v<From>, pi_native_handle>::type
+to_native_handle(From &&from) {
+  return (pi_native_handle)from;
```
Which kind of C++ cast or constructor?

Ralender replied:
it can be either a static_cast or a reinterpret_cast depending on From and it is intended this way

keryell replied:
Add this as a comment

keryell commented on sycl/include/sycl/ext/xilinx/xrt.hpp:
```diff
@@ -0,0 +1,127 @@
+to_native_handle(From &&from) {
+  /// A safe and convenient default is const& with the backend returning without
+  /// copies. a copy will occur on the user's side if needed
+  return (pi_native_handle)std::addressof(from);
```
I am curious about what happens if you really are in the case of `from` is an R-value. returning the address and UB???

Ralender replied:
an r-value reference here will not necessarily cause any UB, if the backend is aware of it (and it should since it chose the type) it can be handled properly and replace a copy by an ownership transfer (std::move).

keryell committed:
Update sycl/plugins/xrt/pi_xrt.cpp

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+  // the pointer will be wrapped into a ref_counted_ref that will increment the
+  // ref_count
+  ref_counted_base() : refCount_(0) {}
+  std::atomic_uint32_t refCount_;
```
Welcome to C++ https://github.com/Xilinx/XRT/pull/6623#discussion_r859212222

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+  /// if the result is 0 the object has been destroyed
+  pi_uint32 decrement_reference_count() noexcept {
+    assert(refCount_ > 0);
+    uint32_t ref_count = --refCount_;
```
Why `pi_uint32` sometimes and `std::uint32_t` sometimes?

Ralender replied:
I turn everything into uint32_t. I don't think it is worth to specify std:: when there is uint32_t is already very clear about what it is.

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+
+/// Will increment ref count if needed by the object
+template <typename T, typename std::enable_if_t<
+                          std::is_base_of_v<ref_counted_base<T>, T>, int> = 0>
```
I am vomiting with the absence of `requires`

Ralender replied:
This is not compiled in C++20 so nothing I can do about it. we have used SFINAE long before requires existed and it works well. I don't think it is that bad.

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+    T *tmp = value;
+    value = other.value;
+    other.value = tmp;
+  }
```
I am puzzled here.
Just swapping values?
`swap` with ADL?

keryell replied:
Why the counters are not swapped?
Comments, comments.

Ralender replied:
this is a ref_counted_ref, the ref-count is inside the ref_counted_base of objects. there is a comment at the start of the object explain what it is.

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+private:
+  /// Storage for the instance
+  static T *&get_self() {
+    static T *self = nullptr;
```
Since you create a `nullptr` there is no static constructor fiasco to fear.
Why not just a static variable?

Ralender replied:
I dislike non-constexpr static data members because you need to add the initialization outside of the class (https://godbolt.org/z/T1s87Pndd).
and in this specific case where the class is template that would mean listing all the possible instantiations of this class.....

keryell replied:
This is no longer the case since C++17 with `inline` members.

keryell replied:
https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:13,endLineNumber:5,positionColumn:13,positionLineNumber:5,selectionStartColumn:13,selectionStartLineNumber:5,startColumn:13,startLineNumber:5),source:'%0Astruct+A+%7B%0A++++inline+static+int+i+%3D+0%3B%0A%7D%3B%0Aint+main()+%7B%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:35.09512000554939,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:g112,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'x86-64+gcc+11.2+(C%2B%2B,+Editor+%231,+Compiler+%231)',t:'0')),k:15.958995626552507,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+11.2+(Compiler+%231)',t:'0')),k:7.979497813276255,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:executor,i:(argsPanelShown:'1',compilationPanelShown:'0',compiler:clang_trunk,compilerOutShown:'0',execArgs:'',execStdin:'',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-std%3Dc%2B%2B20',source:1,stdinPanelShown:'1',tree:'1',wrap:'1'),l:'5',n:'0',o:'Executor+x86-64+clang+(trunk)+(C%2B%2B,+Editor+%231)',t:'0')),k:40.96638655462184,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4

Ralender replied:
oh nice wasn't aware of this.

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+struct cleanup_system {
+  using clenup_item = std::pair<void (*)(void *), void *>;
+  std::vector<clenup_item> cleanup_list;
+  std::unordered_map<void *, unsigned> cleanup_map;
```
```suggestion
  std::vector<cleanup_item> cleanup_list;
  std::unordered_map<void *, unsigned> cleanup_map;
```
also add Doxygen comments on the variables to explain their roles

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+  template <typename T> void add(T *ptr) {
+    unsigned &idx = cleanup_map[ptr];
+    assert(idx == 0);
+    idx = cleanup_list.size();
```
Comment. It took me some time to remember the reference.

Ralender replied:
it is a very common pattern to keep references on map lookup results. to prevent looking up again and again into the map (which is not optimized by the compiler).

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+    unsigned &idx = cleanup_map[ptr];
+    assert(idx == 0);
+    idx = cleanup_list.size();
+    cleanup_list.push_back(clenup_item{[](void *p) { ((T *)p)->~T(); }, ptr});
```
```suggestion
    cleanup_list.emplace_back([](void *p) { ((T *)p)->~T(); }, ptr);
```

Ralender replied:
Why ?

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+
+/// Cannot be used for XRT objects because xrt also has global destructors
+struct cleanup_system {
+  using clenup_item = std::pair<void (*)(void *), void *>;
```
```suggestion
  using cleanup_item = std::pair<void (*)(void *), void *>;
```
but honestly, rather to add this alias, create a `struct` with 2 sensible member names instead of `first` and `second`.

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+template <typename From>
+typename std::enable_if<std::is_reference_v<From>, pi_native_handle>::type
+to_native_handle(From &&from) {
+  return (pi_native_handle)std::addressof(from);
```
No C cast

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+template <typename To>
+typename std::enable_if<std::is_reference_v<To>, To>::type
+from_native_handle(pi_native_handle handle) {
+  return REPRODUCE_ADD_EXTERNAL(std::forward<To>(
```
`std::forward<To>`???

Ralender replied:
yes we the backend can ask to receive an r-value reference to the object if it is asked to keep ownership.

keryell replied:
I see, there is no universal forwarding from some type deduction. It is explicit.

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+private:
+  /// _pi_platform is destroyed during program exit and can happen after XRT's
+  /// global state has been cleanup so we do not hold a counting reference on
+  /// _pi_device. to make it manageable 
```
```suggestion
  /// global state has been cleaned up so we do not hold a counting reference on
  /// _pi_device to make it manageable 
```

Ralender replied:
This comment was outdated.

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+  ref_counted_ref<_pi_device> get_device(unsigned idx) {
+    return devices_[idx];
+  }
+  /// Add a device if it inst't already in the list
```
This is not only what is done.
Comment inside the function too.

Ralender replied:
it is all this function does...

keryell replied:
It also returns the PI device. So it is more like a device factory with a cache or something like that.
This sounds like a good fit for https://www.youtube.com/watch?v=ZDluHz-ybPE ;-)

Ralender replied:
is make_device better ?

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+    void operator()() { function(user_data); }
+  };
+
+  std::vector<ref_counted_ref<_pi_device>> devices_;
```
```suggestion
  // Devices in this context
  std::vector<ref_counted_ref<_pi_device>> devices_;
```

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+  } mem;
+
+  /// Writes cannot be performed on a buffer until the xclbin has been loaded.
+  /// But sycl can request som writes before loading the xclbin so we just
```
```suggestion
  /// But SYCL can request som writes before loading the xclbin so we just
```

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+    }
+    return false;
+  }
+  void map_if_needed(const xrt::device &device, xrt::memory_group grp) {
```
Add an empty line between each function.
Is it not done by `clang-format`?

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+    pending_cmds.exec_queue(); // TODO fix this
+  }
+
+  // TODO: xrt::bo sometimes stay stuck while being deleted. so we dont delete
```
```suggestion
  // TODO: xrt::bo sometimes stay stuck while being deleted. So we do not delete
```

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+  union no_destroy {
+    no_destroy() = default;
+    native_type buffer_ = {};
+    ~no_destroy() {}
```
Why do you need constructors and destructors?

Ralender replied:
because it is not legal to put an non trivially copyable type into an union otherwise.

keryell replied:
How to remember to remove this hack once the `hw_emu` bug is solved?

Ralender replied:
I don't know.

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+  ref_counted_ref<_pi_context> context_;
+  ref_counted_ref<_pi_device> device_;
+
+  workqueue cmds;
```
Comment

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+      : context_{context}, device_{device} {}
+};
+
+typedef void (*pfn_notify)(pi_event event, pi_int32 eventCommandStatus,
```
`using`

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+pi_result getInfo<const char *>(size_t param_value_size, void *param_value,
+                                size_t *param_value_size_ret,
+                                const char *value) {
+  return getInfoArray(strlen(value) + 1, param_value_size, param_value,
```
Ouch `strlen`!

Ralender replied:
strlen isnt that bad, and it is copied from the cuda plugin.

keryell replied:
Sigh!

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+
+/// ------ Error handling, matching OpenCL plugin semantics.
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+namespace pi {
```
C++17

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+
+/// Gets the native XRT handle of a PI device object
+///
+/// \param[in] device The PI device to get the native CUDA object of.
```
Is this copied from CUDA?

Ralender replied:
Yes this was copied from the CUDA plugin and adapted to us

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+pi_result xrt_piextDeviceCreateWithNativeHandle(pi_native_handle handle,
+                                                pi_platform platform,
+                                                pi_device *res) {
+  /// the sycl runtime almost always provide a null platform
```
```suggestion
  /// the SYCL runtime almost always provide a null platform
```

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+  return PI_SUCCESS;
+}
+
+/// Implements a buffer partition in the CUDA backend.
+/// A buffer partition (or a sub-buffer, in OpenCL terms) is simply implemented
+/// as an offset over an existing CUDA allocation.
```
Is this copied from CUDA or OpenCL?

keryell replied:
Remove everywhere.

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+
+pi_result xrt_piEventsWait(pi_uint32 num_events, const pi_event *event_list) {
+  assert_valid_objs(event_list, num_events);
+  /// For now every operation is executed synchronously this ia a noop
```
```suggestion
  /// For now every operation is executed synchronously this is a no op
```

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+  assert_valid_obj(kernel);
+
+  // TODO XRT: analyse if xrt_piKernelSetArg can be called for buffer.
+  REPRODUCE_ADD_BUFFER(arg_value, arg_size);
```
Adding args to buffers???

Ralender replied:
REPRODUCE_ADD_BUFFER only notifies the reproducer system that arg_value is a pointer to a buffer of size arg_size. it is not needed for anything else

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+  buffer->run_when_mapped(command_queue->device_->get(), [=] {
+    void *adjusted_ptr = ((char *)buffer->mem.mapped_ptr) + offset;
+    REPRODUCE_ADD_BUFFER(ptr, size);
+    REPRODUCE_CALL((void)std::memcpy, adjusted_ptr, ptr, size);
```
This sounds super inefficient.

Ralender replied:
XRT doesn't provide any better. options.
but I noticed that for the XRT plugin all data transfer go through the read/write API. where as for the OpenCL plugin most of the calls go through map/unmap. API which we can implement without extra copies. so we could look into this if we want to improve performances

keryell replied:
@stsoe it would be nice to have a way to directly write some random host data to the device `bo`

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+  REPRODUCE_MEMCALL(buffer->get_native(), sync, XCL_BO_SYNC_BO_FROM_DEVICE);
+  void *adjusted_ptr = ((char *)buffer->mem.mapped_ptr) + offset;
+  REPRODUCE_ADD_BUFFER(ptr, size);
+  REPRODUCE_CALL((void)std::memcpy, ptr, adjusted_ptr, size);
```
This sounds super inefficient.

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+  *event = make_ref_counted<_pi_event>().give_externally();
+
+  REPRODUCE_MEMCALL(kernel->run_, start);
+  REPRODUCE_MEMCALL(kernel->run_, wait);
```
Why is this blocking?

Ralender replied:
because we have nothing asynchronous for now.

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+  return PI_SUCCESS;
+}
+
+/// Enqueues a wait on the given CUstream for all events.
```
No CUstream

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+  sycl::detail::pi::unimplemented(__PRETTY_FUNCTION__);
+}
+
+pi_result enuqeue_rect_copy(bool is_read, pi_queue command_queue, pi_mem buffer,
```
```suggestion
pi_result enqueue_rect_copy(bool is_read, pi_queue command_queue, pi_mem buffer,
```

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2312 @@
+
+  _PI_CL(piextKernelSetArgMemObj, xrt_piextKernelSetArgMemObj)
+  _PI_CL(piextKernelSetArgSampler, xrt_piextKernelSetArgSampler)
+  _PI_CL(piTearDown, xrt_piTearDown)
```
I wonder how we can keep in sync all these lines across all the PI...

Ralender replied:
maybe we could, but I don't want to because this list is very useful to navigate through the file.

keryell commented on sycl/plugins/xrt/reproducer.h:
```diff
@@ -0,0 +1,400 @@
+    }
+  }
+  return false;
+}
```
Add comments and design idea.

Ralender replied:
added at the top of the file.

keryell commented on sycl/plugins/xrt/reproducer.h:
```diff
@@ -0,0 +1,400 @@
+  ::detail::reproducer_add_external(std::string(__PRETTY_FUNCTION__) + " " +   \
+                                        __FILE__ + ":" +                       \
+                                        std::to_string(__LINE__),              \
+                                    O)
```
I have a brain overflow arriving to here.

Ralender replied:
I am not sure what I can do about that...

keryell commented on sycl/test/lit.cfg.py:
```diff
@@ -142,6 +143,9 @@
 
 acc_run_substitute=f"env SYCL_DEVICE_FILTER={filter} "
 if xocc != "off":
+    # cleanup the named semaphore in case the previous test didnt cleanup properly
+    # If someone tries to run multiple tests on the same machin this could cause issues.
```
```suggestion
    # Clean up the named semaphore in case the previous test did not clean up properly.
    # If someone tries to run multiple tests on the same machine this could cause issues.
```

keryell commented on sycl/test/lit.cfg.py:
```diff
@@ -142,6 +143,9 @@
 if xocc != "off":
+    # cleanup the named semaphore in case the previous test didnt cleanup properly
+    # If someone tries to run multiple tests on the same machin this could cause issues.
+    os.system("rm -rf /dev/shm/sem.sycl_vxx.py")
```
I guess you assume that this is `rw` by anybody?

Ralender replied:
yeah. now that you say it. I probably need to sudo to cleanup the semaphore...
but haven't seen this cause any issues....

keryell commented on sycl/test/lit.cfg.py:
```diff
@@ -151,11 +155,11 @@
         os.remove(xrt_lock)
     acc_run_substitute="env --unset=XCL_EMULATION_MODE " + acc_run_substitute
     # hw_emu is very slow so it has a higher timeout.
+    acc_run_substitute += "unshare -pc --kill-child "
```
Do not use abbreviated options, specially with such complex commands without comment!

keryell commented on sycl/tools/get_device_count_by_type.cpp:
```diff
@@ -271,6 +275,23 @@ static bool queryHIP(cl_device_type deviceType, cl_uint &deviceCount,
+    deviceCount = 1;
+    return true;
+  } else
+    return true;
```
return true or true?

keryell reviewed (changes requested):
Grand!

Ralender commented:
This commit doesn't address the review, it was just a cleanup commit I thought I had pushed before your review.

Ralender committed:
Apply suggestions from code review

Co-authored-by: Ronan Keryell <ronan@keryell.fr>

Ralender reviewed (commented):
There is a lot of commits because I rebased and forced push.

Ralender commented:
github's handling of force push seems kind of bad...

Commit:
Make pi_xrt.cpp compile under gcc

keryell commented:
I have some problems using this with your example I have simplified to https://github.com/keryell/heterogeneous_examples/blob/main/vector_add/SYCL/vector_add_XRT_interoperability.cpp

```
~/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/clang++ -Wall -std=c++2b -g -g3 -fsycl -fsycl-targets=fpga64_hls_hw_emu -I/opt/xilinx/xrt/include -L/opt/xilinx/xrt/lib -lOpenCL -luuid -lxrt_coreutil -fsycl vector_add_XRT_interoperability.cpp -o vector_add_XRT_interoperability
In file included from vector_add_XRT_interoperability.cpp:2:
/home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/ext/xilinx/xrt.hpp:59:8: error: unknown type name 'device'
inline device make_device<backend::xrt>(
       ^
/home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/ext/xilinx/xrt.hpp:59:27: error: use of undeclared identifier 'backend'; did you mean 'sycl::backend'?
inline device make_device<backend::xrt>(
                          ^
/home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/CL/sycl/backend_types.hpp:21:12: note: 'sycl::backend' declared here
enum class backend : char {
           ^
In file included from vector_add_XRT_interoperability.cpp:2:
/home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/ext/xilinx/xrt.hpp:60:27: error: use of undeclared identifier 'backend'; did you mean 'sycl::backend'?
    const backend_input_t<backend::xrt, device> &BackendObject) {
                          ^
/home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/CL/sycl/backend_types.hpp:21:12: note: 'sycl::backend' declared here
enum class backend : char {
           ^
In file included from vector_add_XRT_interoperability.cpp:2:
/home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/ext/xilinx/xrt.hpp:60:41: error: use of undeclared identifier 'device'
    const backend_input_t<backend::xrt, device> &BackendObject) {
                                        ^
```
etc.

Ralender commented:
I fixed the issue you where having an resolved the merge conflict

Commit:
Revert incorrect namespace change

keryell commented:
I still have a problem:
```
-<2:%>- make vector_add_XRT_interoperability
~/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/clang++ -Wall -std=c++2b -g -g3 -fsycl -fsycl-targets=fpga64_hls_hw_emu -I/opt/xilinx/xrt/include -L/opt/xilinx/xrt/lib -lOpenCL -luuid -lxrt_coreutil -fsycl vector_add_XRT_interoperability.cpp -o vector_add_XRT_interoperability
In file included from vector_add_XRT_interoperability.cpp:2:
In file included from /home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/ext/xilinx/xrt.hpp:11:
/home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/CL/sycl/backend.hpp:54:3: error: implicit instantiation of undefined template 'sycl::detail::BackendInput<sycl::backend::xrt, sycl::device>'
  using input_type = typename detail::BackendInput<Backend, T>::type;
  ^
/home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/CL/sycl/backend.hpp:63:1: note: in instantiation of template type alias 'input_type' requested here
using backend_input_t =
^
/home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/ext/xilinx/xrt.hpp:61:11: note: in instantiation of template type alias 'backend_input_t' requested here
    const backend_input_t<backend::xrt, device> &BackendObject) {
          ^
/home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/CL/sycl/detail/backend_traits.hpp:20:47: note: template is declared here
template <backend Backend, typename T> struct BackendInput;
                                              ^
In file included from vector_add_XRT_interoperability.cpp:2:
In file included from /home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/ext/xilinx/xrt.hpp:11:
/home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/CL/sycl/backend.hpp:54:3: error: implicit instantiation of undefined template 'sycl::detail::BackendInput<sycl::backend::xrt, sycl::device>'
  using input_type = typename detail::BackendInput<Backend, T>::type;
  ^
/home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/CL/sycl/backend.hpp:63:1: note: in instantiation of template type alias 'input_type' requested here
using backend_input_t =
^
/home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/ext/xilinx/xrt.hpp:63:32: note: in instantiation of template type alias 'backend_input_t' requested here
      detail::to_native_handle<backend_input_t<backend::xrt, device>>(
                               ^
/home/rkeryell/Xilinx/Projects/LLVM/worktrees/xilinx/llvm/build/bin/../include/sycl/CL/sycl/detail/backend_traits.hpp:20:47: note: template is declared here
template <backend Backend, typename T> struct BackendInput;
```

Commit:
[XRT][PI] Support asynchronous kernel launches

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -99,27 +99,25 @@ template <typename base> class ref_counted_base {
   ~ref_counted_base() { assert(refCount_ == 0); }
 };
 
+template <typename...> using void_t = void;
```
Why? This is C++17

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -241,6 +239,41 @@ template <typename T> struct unique {
   }
 };
 
+/// Intrusive list base class
```
Seriously?
Use some library like Boost instead?

keryell replied:
OK, as you  want. But could go in some SYCL RT utility somewhere.

keryell commented on buildbot/configure.py:
```diff
@@ -207,6 +212,7 @@ def main():
     parser.add_argument("-t", "--build-type",
                         metavar="BUILD_TYPE", default="Release", help="build type: Debug, Release")
     parser.add_argument("--cuda", action='store_true', help="switch from OpenCL to CUDA")
+    parser.add_argument("--xrt", action='store_true', help="switch from OpenCL to XRT")
```
I guess this is to be also added to the documentation somewhere?

keryell replied:
Here: https://github.com/triSYCL/sycl/blob/sycl/unified/next/sycl/doc/GettingStartedXilinxFPGA.md#compile-the-sycl-compiler

keryell commented on sycl/include/CL/sycl/detail/backend_traits_xrt.hpp:
```diff
@@ -0,0 +1,85 @@
+// sycl::detail::BackendInput and sycl::detail::BackendReturn specialization for
+// the XRT backend.
+//
+// the the supported conversions are:
```
Gnih?

keryell replied:
OK, let's be more factual :-)
```suggestion
// The supported conversions are:
```

keryell commented on sycl/include/CL/sycl/detail/backend_traits_xrt.hpp:
```diff
@@ -0,0 +1,85 @@
+#include <CL/sycl/kernel_bundle.hpp>
+
+/// there is no cyclic dependencies. but the XRT includes path is not configured
+/// by the CMake of the sycl runtime (only the pi_xrt) while compiling this
```
```suggestion
/// There is no cyclic dependencies, but the XRT include path is not configured
/// by the CMake of the SYCL runtime (only the pi_xrt is) while compiling this
```

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2288 @@
+#endif
+
+/// XRT has a bunch of very noisy warnings we need to suppress.
+/// XRT can be found as OpenCL implementation so we suppress warning on both XRT
```
```suggestion
/// XRT might be included also as an OpenCL header implementation, so we suppress warning on both XRT
```

Ralender replied:
its more then might it is happening on sycl41

keryell replied:
Whatever but provide a meaningful comment here

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2288 @@
+#pragma clang diagnostic pop
+
+// Most of functions in the API exists but do not have an implementation so most
+// parameters are not used and it is expected
```
```suggestion
// Most of the functions in the API exist but do not have an implementation, so most
// parameters are not used and it is expected
```

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2288 @@
+    assert(!prev || prev->next == this);
+    assert(!next || next->prev == this);
+  }
+  void insert_next(T *other) {
```
Empty line between all the functions.

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2288 @@
+  }
+};
+
+/// Intrusive list base class
```
```suggestion
/// Intrusive list mix-in
```
?

keryell commented on sycl/plugins/xrt/pi_xrt.cpp:
```diff
@@ -0,0 +1,2288 @@
+
+  /// Writes cannot be performed on a buffer until the xclbin has been loaded.
+  /// But SYCL can request som writes before loading the xclbin so we just
+  /// enqueue them.
```
```suggestion
  /// But SYCL can request some writes before loading the xclbin so we just
  /// enqueue them to process them later.
```

keryell reviewed (changes requested):
Quite an achievement!

keryell reviewed (changes requested):
Quite interesting.

Commit:
Address review comments

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
```diff
@@ -853,7 +853,9 @@ cd $SYCL_HOME
 git clone --branch sycl/unified/master git@github.com:triSYCL/sycl.git llvm
 python $SYCL_HOME/llvm/buildbot/configure.py
-python $SYCL_HOME/llvm/buildbot/compile.py
+python $SYCL_HOME/llvm/buildbot/compile.py --xrt
+# The --xrt is to enable the optional XRT plugin. it is a replacement for the OpenCL plugin.
+# Because the OpencCL implementation in XRT is not maintained anymore
```
```suggestion
# Use --xrt is to enable the optional XRT plugin. This is a replacement for the OpenCL plugin
# because XRT offers more control and expressiveness on the hardware
python $SYCL_HOME/llvm/buildbot/compile.py --xrt
```
to be more factual

keryell reviewed (changes requested):
Almost there!

keryell commented:
@Ralender it would be interesting to see how it works in respect to the presentation from @TApplencourt, @illuhad,... at SYCLcon 2022 _SYCL Concurrency on GPU Platforms: Empirical Measurement_ https://www.youtube.com/watch?v=VqiY-KZXATc&list=PL46sP9LM8GsyHAxj1k7MbWrv5f5SlMpIF&index=19

Ralender commented:
I read through the PR for `xrt::hw_context` and I don't think it is something we need or would benefit to use. and since it is supposed to work as is for existing code I don't think we need to update anything because of it.

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
```diff
@@ -1438,139 +1433,74 @@ the insane! Hopefully...
 
-During the compilation, an temporary directory is created in which all Vitis inputs, commands, outputs and logs are stored.
-By default, this directory is deleted as soon as the compilation ends (even when it fails).
+create a file named reprod.cpp with:
```
```suggestion
Create a file named `reprod.cpp` with:
```

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
```diff
@@ -1438,139 +1433,74 @@ the insane! Hopefully...
+
+int main() {
+
+/// insert the code here
+
+/// Edit the code below to validate the data
+/// int* a_c = name#.data();
+for (int i = 0; i < /*size*/; i++) {
+int res = i + i + 1;
+int val = a_c[i];
+assert(val == res);
+}
+printf("PASS\n");
```
This is not indented?

Ralender replied:
the code automatically generated by the reproducer is not indented so the surrounding was not indented either.
but we can indent it.

keryell replied:
OK.

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
````diff
@@ -1438,139 +1433,74 @@ the insane! Hopefully...
+printf("PASS\n");
+}
+```
+copy the contents of path/to/reprod.out.cpp into reprod.cpp where the comment says so.
````
```suggestion
copy the content of `path/to/reprod.out.cpp` into `reprod.cpp` where the comment says so.
```

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
```diff
@@ -1438,139 +1433,74 @@ the insane! Hopefully...
 
-In order to keep it, set the `SYCL_VXX_KEEP_CLUTTER` environment variable to True.
+the find all the TODOs in the file.
+if your sycl code is simple there should be only 1 TODO which should be replaced by the path to the xclbin see the previous section to get the path to the xclbin
```
```suggestion
Then find all the TODOs in the file.
If your SYCL code is simple, there should be only 1 TODO which should be replaced by the path to the `xclbin` see the previous section to get the path to the `xclbin`.
```

keryell commented on sycl/doc/GettingStartedXilinxFPGA.md:
```diff
@@ -1438,139 +1433,74 @@ the insane! Hopefully...
+if your sycl code is simple there should be only 1 TODO which should be replaced by the path to the xclbin see the previous section to get the path to the xclbin
 
-The compiler will output a similar to
+also edit the end of the main to adapt or remove the validation of data
```
```suggestion
Also edit the end of the `main` to adapt or remove the validation of data
```

keryell commented on sycl/tools/sycl-vxx/bin/sycl_vxx.py:
```diff
@@ -35,6 +36,7 @@
 import subprocess
 import sys
 import tempfile
+import posix_ipc
```
Sort

keryell reviewed (changes requested):
Nice to have some new documentation!
It looks like there are still some comments from me that went unnoticed.

Ralender commented:
> Nice to have some new documentation! It looks like there are still some comments from me that went unnoticed.

yeah I fixed what you asked for but in an other worktree without noticing so the edits you asked are in the pipe PR but not this one. it is fixed it now. and I am looking at new comments.

Commit:
Address Comments

keryell reviewed (approved):
Amazing!
